src/backend/distributed/planner/multi_logical_optimizer.c

```diff
@@ -1611,42 +1694,106 @@ AddTypeConversion(Node *originalAggregate, Node *newExpression)
  * with aggregates in them, this function calls the recursive aggregate walker
  * function to create aggregates for the worker nodes. Also, the function checks
  * if we can push down the limit to worker nodes; and if we can, sets the limit
- * count and sort clause list fields in the new operator node.
+ * count and sort clause list fields in the new operator node. It provides special
+ * treatment for count distinct operator if it is used in repartition subqueries.
+ * Each column in count distinct aggregate is added to target list, and group by
+ * list of worker extended operator.
  */
 static MultiExtendedOp *
 WorkerExtendedOpNode(MultiExtendedOp *originalOpNode)
 {
 	MultiExtendedOp *workerExtendedOpNode = NULL;
+	MultiNode *parentNode = ParentNode((MultiNode *) originalOpNode);
+	MultiNode *childNode = ChildNode((MultiUnaryNode *) originalOpNode);
 	List *targetEntryList = originalOpNode->targetList;
-	List *newTargetEntryList = NIL;
 	ListCell *targetEntryCell = NULL;
+	List *newTargetEntryList = NIL;
+	List *originalGroupClauseList = copyObject(originalOpNode->groupClauseList);
+	List *groupClauseList = NIL;
 	AttrNumber targetProjectionNumber = 1;
+	Index maxSortGroupRefIndex = 0;
+	WorkerAggregateWalkerContext *walkerContext =
+		palloc0(sizeof(WorkerAggregateWalkerContext));
+	walkerContext->repartitionSubquery = false;
+	walkerContext->expressionList = NIL;
+	walkerContext->nextSortGroupRefIndex = 0;
+	walkerContext->groupByList = NIL;
+
+	/* find max of sort group ref index */
+	foreach(targetEntryCell, targetEntryList)
```

Comment:
We use `maxSortGroupRefIndex` only when it's a repartition subquery. So, I think moving this `for` loop and `walkerContext->nextSortGroupRefIndex = maxSortGroupRefIndex + 1;` assignment to the below `if` statement would make it more readable. Because now the readers of the code may think that this is necessary for all kinds of aggregations.